### ISSUE TITLE
Promregator - JSON Logging

### ISSUE THREAD

**Issue by WilliamLHarms** [promregator/promregator#120]
## Environment
I am using the promregator inside of PCF.  

## Summary / Problem Statement
I would like to convert the log output to json.  Looking through the code, it appears you are using org.apache.log4j.Logger;
https://logging.apache.org/log4j/2.x/manual/layouts.html

Would it be safe to add the below json file via -D and set the following property to enable json logging.  (We're trying to balance proper log level output with saving our elastic instance drinking through the firehose)

logging.config=classpath:log4j2.json

```json
{
    "configuration": {
        "appenders": {
            "Console": {
                "name": "STDOUT",
                "JSONLayout": {
                    "compact": "true",
                    "eventEol":"true",
                    "properties":"true",
                    "objectMessageAsJsonObject":"true"
                }
            }
        },
        "loggers": {
            "root": {
                "level": "info",
                "AppenderRef": {
                    "ref": "STDOUT"
                }
            }
        }
    }
}
```

**Comment by WilliamLHarms** [comment]
Well.. My hack didn't work.  Perhaps I'm doingit wrong.

JAVA_OPTS: -Dlogging.config=classpath:$PWD/log4j2.json
^^ In my app manifest 

Also added to the root level of the jar.   I did a cf ssh myApp and saw the file that matched the classpath location.

I also tried file: instead of classpath: no dice.

Patient waiting for JSON logging :)

**Comment by eaglerainbow** [comment]
Hi William,

Welcome to Promregator! 
And thanks for reporting your use case. 

You are right that Promregator is using the **API** of log4j2, but the logging framework really doing the work behind the scenes is logback-classic (as mentioned in the [documentation](https://github.com/promregator/promregator/blob/02487395deb9327936413d4dbdc39c4def050113/docs/documentation.md#logging)). So, your "hack" failed, because you had the wrong configuration file attempt in place.
Keep in mind that logback does not support setting the logging configuration file via command line, but only via the `logback.xml` on the classpath (modulo [various other options](http://logback.qos.ch/manual/configuration.html#auto_configuration), from which I would like to abstract in this case here).
While still approach still provides a huge flexibility, you are still a little doomed with your case, as Promregator does not include the [`logback-contrib` libraries](https://mvnrepository.com/artifact/ch.qos.logback.contrib/logback-json-classic) yet. These are required (besides another hack-ish option), though to get JSON formatting running.

I have added the corresponding support with the current `master` (that will be 0.7.0 and above) and starting with 0.6.2 (see also commits associated to this issue). With these commits I also enhanced the existing documentation, explaining with an example how configuration could look like. 
I briefly tested this case here locally with a test-build of the jar. 

Could you please test-drive this feature and provide feedback?

You say that you are deploying Promregator to PCF. Which type of Promregator deployment are you using, jar or the docker version?
Note that I did not test this with the docker image yet.

**Comment by WilliamLHarms** [comment]
@eaglerainbow  - Thank you so much for pushing a supporting the change.  This will help alot.  Any eta on when the 0.7.0 or 0.6.2 will be posted.  (While I can build my own, I'd prefer the warm fuzzy feeling of downloading from here)

Right now we are dong a simple cf push up to our foundation.  We crack open the jar, add the yml, add the -D's to JAVA_OPTS and that seems to be working.  Please let me know if there's a better way before I write a concourse pipeline to do what I just said :)

Also, if this should be another ticket I can certainly do that, but our promregator is timing out in our off prem PCF platform and sometimes on prem.  After may deep dive sessions with our Pivotal platform engineer (whom pushes for increasing the timeouts to the cloud controller - 2.5s is way to aggressive for his taste)

While the prometheus admin is looking to have us decrease metrics on each scrape. (I don't like this because we loose metric fidelity, going to quantiles instead of histograms)

Here's a few sample errors we are getting:
```
Nov 08 16:07:50 cent-cs-p-0764.dcsg.com java[26381]: 2019-11-08 16:07:50.938 ERROR 26381 --- [     parallel-8] .c.p.c.ReactiveCFPaginatedRequestFetcher : Async retrieval of retrieveAllApplicationIdsInSpace with key acd4809e-1182-4fbf-90d4-208067f3ce92|0bd207e5-4c73-4674-ac34-20d3119ec5ca caused a timeout after 2500ms even though we tried three times

Nov 08 16:06:25 cent-cs-p-0764.dcsg.com java[26381]: 2019-11-08 16:06:25.502 ERROR 26381 --- [     parallel-3] o.c.p.s.ReactiveAppInstanceScanner       : retrieving summary for space id 'd708f5a7-ea31-4b59-ac98-17e990d2853f' resulted in an exception
```

**Comment by eaglerainbow** [comment]
> Thank you so much for pushing a supporting the change. This will help alot. Any eta on when the 0.7.0 or 0.6.2 will be posted. (While I can build my own, I'd prefer the warm fuzzy feeling of downloading from here)

I am currently in the process of building 0.6.2 for you; but I got stuck in a build issue... ETA: any time soon (hopefully within the hour).

**Comment by eaglerainbow** [comment]
> This is the reason why we're trying to run the promregator on platform itself.. We're hoping that we'll squeak enough efficiency out to get below the timeouts.

I don't think that this will change much: Experience shows that the timeouts are (usually) not coming from the network, but from a lame cloud controller. If it was the network differences, then I would first check the network. Having to go over a proxy, though, sometimes might make the difference.

> After may deep dive sessions with our Pivotal platform engineer (whom pushes for increasing the timeouts to the cloud controller - 2.5s is way to aggressive for his taste)

To my mind that depends. How many apps are in the largest space that you try to scrape?

> While the prometheus admin is looking to have us decrease metrics on each scrape. (I don't like this because we loose metric fidelity, going to quantiles instead of histograms)

The timeout error mentioned above is completely unrelated to the amount of metrics that you scrape. Decreasing the number of won't save you from any single (such) timeout message. That is totally sure.

If we want to go deeper into all these matters of timeout et al, I'd suggest to create another issue for that...

**Comment by eaglerainbow** [comment]
> Right now we are ding a simple cf push up to our foundation. We crack open the jar, add the yml, add the -D's to JAVA_OPTS and that seems to be working. Please let me know if there's a better way before I write a concourse pipeline to do what I just said :)

Yea, that's not nice, but up to now I don't have a better idea either. 
Also with the new `logback.xml` you almost don't stand a chance either.

**Comment by WilliamLHarms** [comment]
@eaglerainbow  - thank you for the turn around I'll test it out in our NP envs tomorrow morning.

As for the timeouts, I'll create a new ticket, and yes, for cf push, it's script foo.

**Comment by eaglerainbow** [comment]
I finally managed to get the build issues fixed; 0.6.2 is released.
